optimum/habana/peft/peft_model.py

```diff
@@ -0,0 +1,81 @@
+import warnings
+
+import packaging.version
+import torch
+import transformers
+from peft import PeftType
+
+
+def gaudi_prepare_inputs_for_generation(self, *args, task_ids: torch.Tensor = None, **kwargs):
+    """
+    Copied from PeftModelForCausalLM.prepare_inputs_for_generation: https://github.com/huggingface/peft/blob/v0.9.0/src/peft/peft_model.py#L1156
+    The only differences are:
+    - add token_idx disposal for prompt learning
+    """
+    peft_config = self.active_peft_config
+    model_kwargs = self.base_model_prepare_inputs_for_generation(*args, **kwargs)
+
+    # https://github.com/huggingface/transformers/pull/26681/ introduced new cache format
+    # for some architectures which requires a special fix for prompt tuning etc.
+    # TODO: starting with transformers 4.38, all architectures should support caching.
+    uses_transformers_4_38 = packaging.version.parse(transformers.__version__) >= packaging.version.parse("4.38.0")
+    uses_transformers_4_36 = packaging.version.parse(transformers.__version__) >= packaging.version.parse("4.36.0")
+    transformers_new_cache_archs = ["llama", "mistral", "persimmon", "phi"]
+    uses_cache = uses_transformers_4_38 or (
+        uses_transformers_4_36 and self.base_model.config.model_type in transformers_new_cache_archs
+    )
+
+    if peft_config.peft_type == PeftType.POLY:
+        model_kwargs["task_ids"] = task_ids
+
+    if peft_config.is_prompt_learning:
+        if uses_cache and (model_kwargs["past_key_values"] is not None):
+            # change in the logic of `prepare_inputs_for_generation` makes the below code necessary
+            # In prompt learning methods, past key values are longer when compared to the `input_ids`.
+            # As such only consider the last input ids in the autogressive generation phase.
+            if model_kwargs["past_key_values"][0][0].shape[-2] >= model_kwargs["input_ids"].shape[1]:
+                model_kwargs["input_ids"] = model_kwargs["input_ids"][:, -1:]
+
+        if model_kwargs.get("attention_mask", None) is not None:
+            size = model_kwargs["input_ids"].shape[0], peft_config.num_virtual_tokens
+            prefix_attention_mask = torch.ones(size).to(model_kwargs["input_ids"].device)
+            model_kwargs["attention_mask"] = torch.cat((prefix_attention_mask, model_kwargs["attention_mask"]), dim=1)
+            token_idx = model_kwargs.get("token_idx", None)
+            if token_idx is not None:
+                token_idx = token_idx + peft_config.num_virtual_tokens
```

Comment:
did you check if token_idx_cpu is set for this? if it is, it needs to be updated,right?